### PR TITLE
Dahn list associations + security check. Issues #77 #103

### PR DESCRIPTION
**Pull request by Dahnepfl**
Done : 

### Issue #77 :

- We now have a new item in the admin menu: Fill the associations
- We now list all the associations from the EPFL with an http request (Soon : store in database @njomeau ?)
- Added async tests for the fragment and the url class
- Added function in mainActivity to allow async tests and remove all the sleeps ! :+1: 

    

### About issue #103  :

- Create new superclass `TestWithAdminLogin` to be used for admin fragment test (@Globoxx @njomeau  you can now use this class in tests if you create admin tools)
- Add new menu for admins
- Check in AssociationsGeneratorFragment if the user is an admin. If not, refuse access.
- Create test to check if access to AssociationsGeneratorFragment is well refused to non admin user

### Other changes
- All fragments extends now a supreclass with most commun methods. We got -197 lines in refactoring ! :+1:  See commit https://github.com/njomeau/zuluzulu/pull/107/commits/00e19049d459481a832613b6c42dace673236107
### How to do an async test : 

See for an exemple  `AssociationsUrlHandledTest.java`

```
public class AssociationsUrlHandlerTest {
    @Rule
    public final ActivityTestRule<MainActivity> mActivityRule =
            new ActivityTestRule<>(MainActivity.class);

    private CountingIdlingResource resource;

    private AssociationsUrlHandler object;

    private boolean succes = false;

   @Before
    public void setUp() {

        // Register a new countingIdlingResource
        resource = new CountingIdlingResource("URL Handler test");

        // Register it
        IdlingRegistry.getInstance().register(resource);

        // Increment it. The test will NOT TERMINATE until it's not decremented to zero
        resource.increment();

        // Create the object
        /*
         * In my async task, I give to it the lambda function that will be executed.
         * Here it's a test made function. In the Fragment, it's the function that will get the datas
         * and print them in the view.
         *
         * This test function handler() will be executed when the action is finished
         */
        this.object = new AssociationsUrlHandler(this::handler);
    }

    @Test
    public void worksWithGoodURL()
    {
        // Execute our async task
        object.execute(AssociationsGeneratorFragment.EPFL_URL); // Test with real URL

        // We need to open a View for the test to start waiting...
        Utility.openMenu();

        // wait....

        // stop waiting on resource.decrement()   OR timeout !

        assertTrue(succes);
    }

    /**
     * This function will get the datas
     *
     * @param result arrayList
     * @return Void
     */
    private Void handler(ArrayList<String> result) {
        // Do any logic. Here we want result not to be null
        succes = result != null;

        // Free the resource with decrement. Now the test stop waiting !
        resource.decrement();

        return null;
    }

}
```

I have implemented a similar incremented and decremented function in the main activity 
I have done it as said in https://developer.android.com/training/testing/espresso/idling-resource#integrate-recommended-approach

**So async test in fragments are WAY EASYER :**

So now before a async task : 
```
        // Send increment to wait async execution in test
        mListener.onFragmentInteraction(MainActivity.INCREMENT, true);
```

In the function that will handle the post execution of the async task :
        
```
// Tell tests the async execution is finished
        mListener.onFragmentInteraction(MainActivity.DECREMENT, true);
```

You just need to register the resource in the Tests with

 `IdlingRegistry.getInstance().register(mActivityRule.getActivity().getCountingIdlingResource());`

Maybe I will add this in the Utility class...

If it's badly explained, feel free to ask

### CODE REVIEW

**Comment by Dahnepfl** [comment]
Please note that you can run multiple async tast with few increment and few decrement.

Test will wait until it's decremented to zero

**Comment by Dahnepfl** [comment]
Thank you ! I will look a this before merging

**Comment by Globoxx** [comment]
> Thank you !
> 
> It's because I put the associations data in an arraylist. But I can maybe change this. I will look a this before merging

@Dahnepfl Yes but I mean, like that you can only call your functions with an ArrayList. You have nothing to change to your implementation, just allowing a more generic type of data as input.
Unless you use specific properties of the arraylist in your methods.

**Comment by Dahnepfl** [comment]
> > Thank you !
> > It's because I put the associations data in an arraylist. But I can maybe change this. I will look a this before merging
> 
> @Dahnepfl Yes but I mean, like that you can only call your functions with an ArrayList. You have nothing to change to your implementation, just allowing a more generic type of data as input.
> Unless you use specific properties of the arraylist in your methods.

Done !

**Comment by Dahnepfl** [comment]
New addition : 

- Create new superclass `TestWithAdminLogin` to be used for admin fragment test (@Globoxx @njomeau  you can now use this class in tests if you create admin tools)
- Add new menu for admins
- Check in AssociationsGeneratorFragment if the user is an admin. If not, refuse access.
- Create test to check if access to AssociationsGeneratorFragment is well refused

File: app/src/main/java/ch/epfl/sweng/zuluzulu/Fragments/SuperFragment.java [hunk 1]
```diff
@@ -0,0 +1,28 @@
+package ch.epfl.sweng.zuluzulu.Fragments;
+
+import android.content.Context;
+import android.support.v4.app.Fragment;
+
+import ch.epfl.sweng.zuluzulu.OnFragmentInteractionListener;
+
+public class SuperFragment extends Fragment {
+    protected OnFragmentInteractionListener mListener;
+
```
Comment (Dahnepfl) [inline app/src/main/java/ch/epfl/sweng/zuluzulu/Fragments/SuperFragment.java]: @Globoxx @njomeau 
I created a super fragment class. It implements the listener. So we don't have to do it on each fragment and we can remove methods onAttach and onDetach on each fragment. So -10 lines per fragment.

Is it ok for you ?

File: app/src/main/java/ch/epfl/sweng/zuluzulu/Fragments/SuperFragment.java [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+
+    @Override
+    public void onAttach(Context context) {
+        super.onAttach(context);
+        if (context instanceof OnFragmentInteractionListener) {
+            mListener = (OnFragmentInteractionListener) context;
```
Comment (Dahnepfl) [inline app/src/main/java/ch/epfl/sweng/zuluzulu/Fragments/SuperFragment.java]: Overrided class

File: app/src/main/java/ch/epfl/sweng/zuluzulu/Fragments/SuperFragment.java [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+        } else {
+            throw new RuntimeException(context.toString()
+                    + " must implement OnFragmentInteractionListener");
+        }
+    }
+
+    @Override
+    public void onDetach() {
+        super.onDetach();
+        mListener = null;
```
Comment (Dahnepfl) [inline app/src/main/java/ch/epfl/sweng/zuluzulu/Fragments/SuperFragment.java]: Overrided class

File: app/src/main/java/ch/epfl/sweng/zuluzulu/Fragments/AssociationsGeneratorFragment.java [hunk 4]
```diff
@@ -95,21 +94,4 @@ public View onCreateView(LayoutInflater inflater, ViewGroup container,
         // Inflate the layout for this fragment
         return inflater.inflate(R.layout.fragment_associations_generator, container, false);
     }
-
-    @Override
-    public void onAttach(Context context) {
-        super.onAttach(context);
-        if (context instanceof OnFragmentInteractionListener) {
```
Comment (Dahnepfl) [inline app/src/main/java/ch/epfl/sweng/zuluzulu/Fragments/AssociationsGeneratorFragment.java]: Code we can remove on EACH FRAGMENT

**Comment by Dahnepfl** [comment]
Hi @betty 

Here is my work of the previous week. There is also pr #117 #106.

**Pull request merged**